### PR TITLE
Replicate ammonia nitrogen fraction fix in dev

### PR DESCRIPTION
**Pull request by KFosterReed**
Fixes mathematical errors in `NutrientRequestResults.__add__`.

### What
The equations that combine the nitrogen values when adding two `NutrientRequestResults` objects were incorrect, due to some confusing fraction math. They've been refactored to a more logical structure, which also fixed the math errors.

### Test plan
* Added three separate unit tests to check that nitrogen, phosphorus, and dry matter are correctly added when adding two `NutrientRequestResults` objects.

### Input Changes
None

### Output Changes
None

### CODE REVIEW

**Comment by morrowcj** [comment]
@KFosterReed, As I am writing the tests, I've become a bit confused by a few things in the ``__add__`` method:
* When ``combined_inorganic_nitrogen <= 0``, then: 
  ```python
  combined_ammonium_nitrogen_fraction == self.ammonium_nitrogen_fraction
  ``` 
* and similarly when ``combined_total_nitrogen <= 0``, then:
  ```python
  combined_organic_nitrogen_fraction = self.organic_nitrogen_fraction
  combined_inorganic_nitrogen_fraction = self.inorganic_nitrogen_fraction
  combined_ammonium_nitrogen_fraction = self.ammonium_nitrogen_fraction
  ```
but I'm having trouble understanding why. 

Similarly, what happens when one of the fractions (e.g., `inorganic_nitrogen_fraction`) are zero in one or more of the requests? You get a divide by zero errors currently, but perhaps that doesn't matter if they are never able to be zero in practice.

**Comment by morrowcj** [comment]
@KFosterReed, I think this PR is ready for review. The tests are all passing, so I think the logic is good and the maths correct!

There are a few `#TODOs` in the test file that flag the areas that I wanted clarification about (described in my previous comment). We can remove them prior to merging.

I've also edited the main PR comment to describe my tests.

**Comment by KFosterReed** [comment]
> @KFosterReed, As I am writing the tests, I've become a bit confused by a few things in the `__add__` method:
> 
> * When `combined_inorganic_nitrogen <= 0`, then:
>   ```python
>   combined_ammonium_nitrogen_fraction == self.ammonium_nitrogen_fraction
>   ```
> * and similarly when `combined_total_nitrogen <= 0`, then:
>   ```python
>   combined_organic_nitrogen_fraction = self.organic_nitrogen_fraction
>   combined_inorganic_nitrogen_fraction = self.inorganic_nitrogen_fraction
>   combined_ammonium_nitrogen_fraction = self.ammonium_nitrogen_fraction
>   ```
> 
> but I'm having trouble understanding why.
> 
> Similarly, what happens when one of the fractions (e.g., `inorganic_nitrogen_fraction`) are zero in one or more of the requests? You get a divide by zero errors currently, but perhaps that doesn't matter if they are never able to be zero in practice.

This is a great question Clay. I might bring @matthew7838 in because I think he was the original author of this method but I believe that the intention behind these `<= 0` methods were to catch edge cases that shouldn't really happen. As you point out though, as you point out, it seems like this is not currently doing that anyway.

**Comment by matthew7838** [comment]
> doesn't matter if they are never able to be zero in practice.

So, actually, I didn't touch this part of the code while doing the manure refresh connection. However, from my understanding of this part of the code, yes, @KFosterReed was right, the <= 0 is merely a code-level logic implementation, nothing specific.

As for the second part of @morrowcj's question, again, since these checks are merely code-level, we don't care about the science aspect of whether being 0 makes sense or not. As long as the failure entry points (combined_total_inorganic_nitrogen, combined_total_nitrogen, and combined_total_phosphorus) are not 0, we successfully avoid zero-division errors.

However, @morrowcj, I don't understand how inorganic_nitrogen_fraction being 0 can lead to a zero-division error under the current code logic. Can you elaborate a bit more on the case you have?

**Comment by morrowcj** [comment]
> So, actually, I didn't touch this part of the code while doing the manure refresh connection. However, from my understanding of this part of the code, yes, @KFosterReed was right, the <= 0 is merely a code-level logic implementation, nothing specific.

OK, that makes sense. I do wonder if there's a use case where two `NutrientRequest`s both have no nitrogen (but e.g., some phosphorus etc.). In that case, would it make more sense to set all of the nitrogen fractions to 0, since none of total mass is comprised of nitrogen?

> However, @morrowcj, I don't understand how inorganic_nitrogen_fraction being 0 can lead to a zero-division error under the current code logic. Can you elaborate a bit more on the case you have?

Yeah, after double checking on this, it was my error.

File: RUFAS/data_structures/manure_to_crop_soil_connection.py
Comment (morrowcj) [inline RUFAS/data_structures/manure_to_crop_soil_connection.py:182]: Should we add something to the doc strings to describe the logic?

File: tests/test_data_structures/test_manure_nutrients.py [hunk 1]
```diff
@@ -4,9 +4,162 @@
 from pytest import approx, mark
 
 from RUFAS.data_structures.manure_nutrients import ManureNutrients
+from RUFAS.data_structures.manure_to_crop_soil_connection import NutrientRequestResults
 from RUFAS.data_structures.manure_types import ManureType
 
 
+@mark.parametrize("nitrogen_one", [100, 0])
+@mark.parametrize("nitrogen_two", [125, 0])
+@mark.parametrize("inorganic_frac_one", [0.5, 0.01, 0.99])  # TODO: how to handle zero?
```
Comment (morrowcj) [inline tests/test_data_structures/test_manure_nutrients.py]: This TODO reminds me that I was wondering if we needed to consider and account for the possible edge case of these fractions (and their respective quantities) being 0. This might occur if, for example, all of the nitrogen occurs as a single form. If That's not a concern, we can delete this comment.

```suggestion
@mark.parametrize("inorganic_frac_one", [0.5, 0.01, 0.99])
```

File: tests/test_data_structures/test_manure_nutrients.py [hunk 2]
```diff
@@ -4,9 +4,162 @@
...
+        organic_nitrogen_fraction=organic_frac_two,
+        inorganic_nitrogen_fraction=inorganic_frac_two,
+        ammonium_nitrogen_fraction=ammonium_frac_two,
+    )
+    # Act
+
+    # Expected values
+    expected_nitrogen = nitrogen_one + nitrogen_two
+
+    if expected_nitrogen <= 0:  # both zero TODO: I'm confused by this.
```
Comment (morrowcj) [inline tests/test_data_structures/test_manure_nutrients.py]: I think that my thought with this comment is that, if the expected nitrogen is 0, shouldn't all these fractions be 0 also, rather than their default (which may or may not be zero).

File: tests/test_data_structures/test_manure_nutrients.py [hunk 3]
```diff
@@ -4,9 +4,162 @@
...
+    # Assert
+    assert combined_request.nitrogen == pytest.approx(expected_nitrogen)
+    assert combined_request.inorganic_nitrogen_fraction == pytest.approx(expected_inorganic_frac)
+    assert combined_request.organic_nitrogen_fraction == pytest.approx(expected_organic_frac)
+    assert combined_request.ammonium_nitrogen_fraction == pytest.approx(expected_ammonium_frac)
+
+
+@mark.parametrize("phosphorus_one", [100, 0])
+@mark.parametrize("phosphorus_two", [125, 0])
+@mark.parametrize("inorganic_frac_one", [0.5, 0.01, 0.99])  # TODO: how to handle zero?
```
Comment (morrowcj) [inline tests/test_data_structures/test_manure_nutrients.py]: Same as above.

```suggestion
@mark.parametrize("inorganic_frac_one", [0.5, 0.01, 0.99])
```

File: tests/test_data_structures/test_manure_nutrients.py [hunk 4]
```diff
@@ -4,9 +4,162 @@
...
+
+    # Assert
+    assert combined_request.phosphorus == pytest.approx(expected_phosphorus)
+    assert combined_request.inorganic_phosphorus_fraction == pytest.approx(expected_inorganic_fraction)
+    assert combined_request.organic_phosphorus_fraction == pytest.approx(expected_organic_fraction)
+
+
+@mark.parametrize("mass_one", [100, 0])
+@mark.parametrize("mass_two", [125, 0])
+@mark.parametrize("dry_frac_one", [0.5, 0.01, 0.99])  # TODO: how to handle zero?
```
Comment (morrowcj) [inline tests/test_data_structures/test_manure_nutrients.py]:
```suggestion
@mark.parametrize("dry_frac_one", [0.5, 0.01, 0.99]) 
```

**Review by morrowcj — APPROVED** [review]
Looks good to me. Left a few comments about the TODOs that I'd left.

File: tests/test_data_structures/test_manure_nutrients.py [hunk 5]
```diff
@@ -4,9 +4,162 @@
...
+
+    if expected_phosphorus <= 0:  # TODO: I'm confused by this
+        expected_inorganic_fraction = inorganic_frac_one
+        expected_organic_fraction = organic_frac_one
+    else:
+        inorganic_one = inorganic_frac_one * phosphorus_one
+        inorganic_two = inorganic_frac_two * phosphorus_two
+        expected_inorganic_fraction = (inorganic_one + inorganic_two) / expected_phosphorus
+
+        expected_organic_fraction = 1 - expected_inorganic_fraction  # works here because of our setup above.
```
Comment (matthew7838) [inline tests/test_data_structures/test_manure_nutrients.py]:
```suggestion
        expected_organic_fraction = 1 - expected_inorganic_fraction
```

**Review by matthew7838 — APPROVED** [review]
LGTM, this PR is missing changelog entry, is that something to be done in the other PR? Be sure to remove all in-line comments before merging.
